### PR TITLE
feat(cli): JSON reporter

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

This PR exposes (again) the `--json` option, and it adds a new `--json-pretty`.

The `--json` option will print everything in JSON format (diagnostics and summary). 

The `--json-pretty`, together with `--json`, will print said JSON formatted.

I also added a warning because the fields' names are weird, and I think they should be reviewed. 

## Test Plan

I added new tests

### CODE REVIEW

**Comment by Conaclos** [comment]
This reporter is intended for machines not human? Thus, I am not sure if we need `--json-pretty`. We could first introduce only `--json` and see users reaction.

File: CHANGELOG.md [hunk 1]
```diff
@@ -94,6 +94,11 @@ our [guidelines for writing a good changelog entry](https://github.com/biomejs/b
 
   Contributed by @Conaclos
 
+- Added two new options to customise the emitted output of the CLI: `--json` and `--json-pretty`. With `--json`, the diagnostics and the
+  summary will be printed in the **terminal** in JSON format. Using `--json-pretty` _together_ with `--json` will output the same output, but formatted.
```
Comment (arendjr) [inline CHANGELOG.md:98]: Why do they need to be combined? The fact it requires so much emphasis seems to recognize it's a potential gotcha. Can't we make it more friendly by allowing `--json-pretty` to be used on its own?

Comment (ematipico) [reply CHANGELOG.md:98]: I thought about it, but this combination would require a check if both are passed and raise an error if they are.

I'm open to better names or options :)

Comment (arendjr) [reply CHANGELOG.md:98]: If both are specified, we can just let `--json-pretty` win, no?

File: crates/biome_cli/src/reporter/mod.rs [hunk 2]
```diff
@@ -28,58 +30,24 @@ pub struct TraversalSummary {
...
+        execution: &Execution,
+        summary: TraversalSummary,
     ) -> io::Result<()>;
 
     /// Writes a diagnostics
-    fn report_diagnostics(&mut self, payload: &DiagnosticsPayload) -> io::Result<()>;
-}
-
-/// This function reports the result of a traversal
-pub(crate) fn report<R, V>(
```
Comment (ematipico) [inline crates/biome_cli/src/reporter/mod.rs:50]: This function was causing compiler issues, it's not needed. The error was around `Reporter` not having a size that could not be known at compile time

File: crates/biome_cli/src/reporter/terminal.rs [hunk 3]
```diff
@@ -1,69 +1,22 @@
...
 
-#[derive(Default)]
-pub(crate) struct ConsoleReporterBuilder<'a> {
-    cli_options: Option<&'a CliOptions>,
-    execution: Option<&'a Execution>,
-    summary: Option<&'a TraversalSummary>,
-    diagnostics: Vec<Error>,
-}
-
-impl<'a> ConsoleReporterBuilder<'a> {
```
Comment (ematipico) [inline crates/biome_cli/src/reporter/terminal.rs:25]: this is not needed anymore, we create the reporter upstream

File: crates/biome_cli/src/reporter/terminal.rs [hunk 4]
```diff
@@ -90,13 +43,17 @@ impl<'a> ReporterVisitor for ConsoleReporterVisitor<'a> {
...
         });
 
         Ok(())
     }
 
-    fn report_diagnostics(&mut self, diagnostics_payload: &DiagnosticsPayload) -> io::Result<()> {
+    fn report_diagnostics(
+        &mut self,
+        execution: &Execution,
+        diagnostics_payload: DiagnosticsPayload,
```
Comment (ematipico) [inline crates/biome_cli/src/reporter/terminal.rs:55]: We must take ownership of the diagnostics because we need to convert them to serde diagnostics, and the `Error` type can't be cloned.

File: crates/biome_cli/src/execute/mod.rs [hunk 5]
```diff
@@ -362,22 +369,87 @@ pub fn execute_mode(
...
+                };
+                let mut buffer = JsonReporterVisitor::new(summary_result);
+                reporter.write(&mut buffer)?;
+                if pretty {
+                    let content = serde_json::to_string(&buffer).map_err(|error| {
+                        CliDiagnostic::Report(ReportDiagnostic::Serialisation(
+                            SerdeJsonError::from(error),
+                        ))
+                    })?;
+                    let report_file = BiomePath::new("_report_output.json");
```
Comment (ematipico) [inline crates/biome_cli/src/execute/mod.rs:414]: I wanted to format the JSON file using the same options that the user defined, that's why I used the workspace. Although, this is a hazard and causes issues if a user has a `_report_output.json` file. 

Ideally, in the future, we want to create a virtual file that doesn't conflict with the files of the user.

File: CHANGELOG.md [hunk 6]
```diff
@@ -94,6 +94,67 @@ our [guidelines for writing a good changelog entry](https://github.com/biomejs/b
...
```
Comment (arendjr) [inline CHANGELOG.md:98]:
```suggestion
  summary will be printed in the **terminal** in JSON format. `--json-pretty` does the same, but also runs the formatter over the generated JSON. If both are supplied, `--json-pretty` takes precedence.
```

File: crates/biome_cli/src/cli_options.rs [hunk 7]
```diff
@@ -46,10 +46,14 @@ pub struct CliOptions {
     #[bpaf(long("error-on-warnings"), switch)]
     pub error_on_warnings: bool,
 
-    /// Reports information using the JSON format
-    #[bpaf(long("json"), switch, hide_usage, hide)]
+    /// It prints to terminal diagnostics and summary in JSON format
```
Comment (arendjr) [inline crates/biome_cli/src/cli_options.rs:49]: I find the old documentation line to be more understandable. "Report diagnostics and summary in JSON format" would also be fine.

I see three problems with the new documentation line:
- "It prints..." Who is "it"? The command-line argument isn't an actor, it's a directive. So it can't be the subject in this sentence. The other documentation lines avoid this confusion by leaving the subject out of the description. Simply saying "Print" or "Report" is more clear here. (This is slightly different from documenting functions or methods, because they can be seen as actors, though even there it's often fine to omit the subject for brevity.)
- "prints to terminal" Why is the terminal relevant? If the argument is not supplied, it also prints to terminal, right? So mentioning it here distracts from the message you're trying to convey.
- "terminal diagnostics" Terminal diagnostics are diagnostics about the terminal, or diagnostics printed in the terminal. As such, they can be read together as a single object. But of course, that's not how they're intended in this sentence. The sentence intends to say "prints to terminal, diagnostics and summary". Adding the comma would take the flow out of the sentence however, but omitting it makes the sentence hard to parse for humans, since a human reader effectively has to back-track when they realize that interpreting "terminal diagnostics" as a single object will lead to a grammatical mistake near the end of the sentence. Fortunately this is also fixed by removing the mention of the terminal, but if it had been vital to mention the terminal, this construct would've been better split into two sentences.

Not mentioning these to pick on you, of course. Hopefully they can help us all to write clearer documentation lines!

File: crates/biome_cli/src/diagnostics.rs [hunk 8]
```diff
@@ -262,6 +264,12 @@ impl DeprecatedArgument {
     }
 }
 
+#[derive(Debug, Diagnostic)]
+pub enum ReportDiagnostic {
+    /// Emitted when trying to serialise the report
+    Serialisation(SerdeJsonError),
```
Comment (arendjr) [inline crates/biome_cli/src/diagnostics.rs:270]:
```suggestion
    Serialization(SerdeJsonError),
```

Writing Serialisation with an "s" is UK English. I prefer UK English myself (I use it to write my novels!), but for programming-related matters, US English is the gold standard.

**Review by arendjr — APPROVED** [review]
Nice feature! Only left some nits :)

File: crates/biome_cli/src/cli_options.rs (hunk 7 above)
Comment (ematipico) [reply crates/biome_cli/src/cli_options.rs]: It's OK to provide suggestions :) I am not good at writing and I am **well** aware of that 🤣

File: crates/biome_cli/src/execute/mod.rs (hunk 5 above)
Comment (Conaclos) [reply crates/biome_cli/src/execute/mod.rs:414]: We could also add `biome` in the filename?

**Comment by Conaclos** [comment]
I forgot to add this comment:

Could not use an option `--reporter=<reporter>` instead of specific flags?
We could have `--reporter=json` and `--reporter=json-pretty`.
This could avoid the addition of more flags once we integrate more reporters.

**Comment by ematipico** [comment]
Sure, no problem

**Comment by Conaclos** [comment]
> Sure, no problem

Do you plan to make the change?
